src/basic/os-util.c

```diff
@@ -67,6 +70,70 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
+
+                        r = -ENOENT;
+                        struct dirent *de;
+                        FOREACH_DIRENT(de, extension_release_dir, return -errno) {
+                                int k;
+
+                                if (!IN_SET(de->d_type, DT_REG, DT_UNKNOWN))
+                                        continue;
+
+                                const char *image_name = startswith(de->d_name, "extension-release.");
+                                if (!image_name)
+                                        continue;
+
+                                if (!image_name_is_valid(image_name))
+                                        continue;
+
+                                /* We already chased the directory, and checked that
+                                 * this is a real file, so we shouldn't fail to open it. */
+                                _cleanup_close_ int extension_release_fd = openat(dirfd(extension_release_dir), de->d_name, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+                                if (extension_release_fd < 0)
+                                        return log_debug_errno(errno, "Failed to open extension-release file %s/%s: %m", extension_release_dir_path, de->d_name);
+
+                                /* No xattr or cannot parse it? Then skip this. */
+                                _cleanup_free_ char *extension_release_xattr = NULL;
+                                k = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+                                if (k < 0 && !ERRNO_IS_NOT_SUPPORTED(k) && k != -ENODATA)
+                                        log_debug_errno(k, "Failed to read 'user.extension-release.strict' extended attribute from extension-release file %s/%s: %m", extension_release_dir_path, de->d_name);
+                                if (k < 0)
+                                        continue;
+
+                                /* Explicitly set to request strict matching? Skip it. */
+                                k = parse_boolean(extension_release_xattr);
+                                if (k < 0)
+                                        log_debug_errno(k, "Failed to parse 'user.extension-release.strict' extended attribute value from extension-release file %s/%s: %m", extension_release_dir_path, de->d_name);
+                                if (k < 0 || k > 0)
+                                        continue;
+
+                                r = 0; /* Found it! */
+
+                                if (ret_fd)
+                                        fd = TAKE_FD(extension_release_fd);
+
+                                if (ret_path) {
+                                        q = path_join(extension_release_dir_path, de->d_name);
+                                        if (!q)
+                                                return -ENOMEM;
+                                }
+
+                                break;
```

Comment:
instead of leaving the loop here, I think we should go on, and check if there's another matching file. If so we should fail  (error ENOTUNIQ?). i.e. i'd be a bit strict here, even in the relaxed mode: insist that there's only a single extension file in the image.

Reply:
Do you want to parse the entire directory even if there's an exact match then? TBH I'm not sure it's worth it, with the exact matching or the xattr... also I'd need to skip .p7s, as in some images I use a blanket fs-verity on everything

Reply:
if there's an exact match (i.e. one that includes the image name), then we shouldn't care what else is in the extensions.d/. But if we do that generic fallback, where we allow "wildcard" matches, as your PR implements it, then we should iterate through the dir to ensure there are no ambiguities. 

The way I see it: it should be totally OK to build an image that can act as multiple different extensions, if that's what people do. But if so, we need to know under which  extension to use it, i.e. we need the extension name as input to find the right of the extension release file. In that case it is safe to ignore any other release files, because it's clear what is meant. However, if we do the generic match then we really should expose systematic, defined behaviour, and remove all ambiguities, so that it cannot happen that you attach it to day and another time tomorrow, and we consider it a different extension. if you follow what i mean

Reply:
ok, if there's another match with the xattr I'll raise an error. PTAL